src/charm.py

```diff
@@ -326,19 +332,68 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
 
         self._initialise_replica_set(event)
 
-    def _on_update_status(self, _):
+    def _on_update_status(self, event):
         # cannot have both legacy and new relations since they have different auth requirements
         if self.client_relations._get_users_from_relations(
             None, rel="obsolete"
         ) and self.client_relations._get_users_from_relations(None):
             self.unit.status = BlockedStatus("cannot have both legacy and new relations")
             return
 
-        # if unit is primary then update status
-        if self._primary == self.unit.name:
-            self.unit.status = ActiveStatus("Replica set primary")
-        else:
-            self.unit.status = ActiveStatus()
+        # no need to report on replica set status until initialised
+        if "db_initialised" not in self.app_peer_data:
+            return
+
+        # leader should periodically handle configuring the replica set. Incidents such as network
+        # cuts can lead to new IP addresses and therefore will require a reconfigure. Especially
+        # in the case that the leader a change in IP address it will not receive a relation event.
+        if self.unit.is_leader():
+            self._handle_reconfigure(event)
+
+        # update the units status based on it's replica set status.
+        with MongoDBConnection(self.mongodb_config) as mongo:
+            replset_status = mongo.get_replset_status()
+
+            if self._unit_ip(self.unit) not in replset_status:
+                self.unit.status = WaitingStatus("Member being added..")
+                return
+
+            replica_status = replset_status[self._unit_ip(self.unit)]
+
+            if replica_status == "PRIMARY":
+                self.unit.status = ActiveStatus("Replica set primary")
+            elif replica_status == "SECONDARY":
+                self.unit.status = ActiveStatus("Replica set secondary")
+            elif replica_status in ["STARTUP", "STARTUP2", "ROLLBACK", "RECOVERING"]:
+                self.unit.status = WaitingStatus("Member is syncing..")
+            elif replica_status == "REMOVED":
+                self.unit.status = WaitingStatus("Member is removing..")
+            else:
+                self.unit.status = BlockedStatus(replica_status)
+
+    def _handle_reconfigure(self, event):
+        """Reconfigures the replica set if necessary.
+
+        Removes any mongod hosts that are no longer present in the replica set or adds hosts that
+        should exist in the replica set. This function is meant to be called periodically by the
+        leader in the update status hook to perform any necessary cluster healing.
+        """
+        if not self.unit.is_leader():
+            logger.debug("only the leader can perform reconfigurations to the replica set.")
```

Comment:
Do you want to `return` here if non-leader?
I didn't see `self.update_app_relation_data()` check for leader, could have missed it.